crates/wasi-common/src/sandboxed_tty_writer.rs

```diff
@@ -0,0 +1,177 @@
+                '\u{0002}' => '␂',
+                '\u{0003}' => '␃',
+                '\u{0004}' => '␄',
+                '\u{0005}' => '␅',
+                '\u{0006}' => '␆',
+                '\u{0007}' => '␇',
+                '\u{0008}' => '␈',
+                '\u{0009}' => '\t',
+                '\u{000A}' => '\n',
+                '\u{000B}' => '␋',
+                '\u{000C}' => '␌',
+                '\u{000D}' => '\r',
+                '\u{000E}' => '␎',
+                '\u{000F}' => '␏',
+                '\u{0010}' => '␐',
+                '\u{0011}' => '␑',
+                '\u{0012}' => '␒',
+                '\u{0013}' => '␓',
+                '\u{0014}' => '␔',
+                '\u{0015}' => '␕',
+                '\u{0016}' => '␖',
+                '\u{0017}' => '␗',
+                '\u{0018}' => '␘',
+                '\u{0019}' => '␙',
+                '\u{001A}' => '␚',
+                '\u{001B}' => '␛',
+                '\u{001C}' => '␜',
+                '\u{001D}' => '␝',
+                '\u{001E}' => '␞',
+                '\u{001F}' => '␟',
+                '\u{007F}' => '␡',
+                x if x.is_control() => '�',
+                x => x,
+            }
+            .encode_utf8(&mut self.scratch)
+            .as_bytes(),
+        )?;
+
+        Ok(())
+    }
+
+    /// Write a string to the output.
+    pub(crate) fn write_str(&mut self, s: &str) -> Result<usize> {
+        let mut result = 0;
+
+        for c in s.chars() {
+            self.write_char(c)?;
+            // Note that use the encoding length of the given char, rather than
+            // how many bytes we actually wrote, because our users don't know about
+            // what's really being written.
+            result += c.len_utf8();
+        }
+
+        Ok(result)
+    }
+}
+
+impl<'writer, Writer> Write for SandboxedTTYWriter<'writer, Writer>
+where
+    Writer: Write,
+{
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        let mut input = buf;
+        let mut result = 0;
+
+        // Decode the string without heap-allocating it. See the example here
+        // for more details:
+        // https://doc.rust-lang.org/std/str/struct.Utf8Error.html#examples
+        loop {
+            match ::std::str::from_utf8(input) {
+                Ok(valid) => {
+                    result += self.write_str(valid)?;
+                    break;
+                }
+                Err(error) => {
+                    let (valid, after_valid) = input.split_at(error.valid_up_to());
+                    result += self.write_str(unsafe { std::str::from_utf8_unchecked(valid) })?;
+                    self.write_char('�')?;
+
+                    if let Some(invalid_sequence_length) = error.error_len() {
+                        result += invalid_sequence_length;
+                        input = &after_valid[invalid_sequence_length..];
+                    } else {
+                        result += after_valid.len();
```

Comment:
Hmm, I might be wrong here but it looks to me as though we increment `result` by the same value in both branches, whereas I guess the first branch should increment it by either 1 or the actual length of the replacement char?

Reply:
`result` is the value we return back to the application to tell it how many bytes we wrote, and the application doesn't know that we used replacement characters instead of the bytes it gave us.

In the true arm of this if, we increment the result by the length of the invalid sequence, telling the application that we've "written" those bytes, and then resume normal writing after that point.

In the false arm of this if, the end of the buffer has been reached, so we just increment the result by the number of bytes until the end of the buffer.

I'll add some comments about this.